makefiles/features.modules.mk

```diff
@@ -0,0 +1,37 @@
+# Add modules implementing used features
+#
+# This is done after the regular dependency resolution in Makefile.dep, as
+# feature resolution depends on the used modules. As these modules however have
+# no dependencies(except for periph_common), no dependency resolution is needed.
+
+PERIPH_FEATURES := $(filter periph_%,$(FEATURES_USED))
+# all periph features correspond to a periph submodule
+# FEATURES_USED is defined in Makefile.features
+USEMODULE += $(PERIPH_FEATURES)
+
+# select cpu_check_address pseudomodule if the corresponding feature is used
+USEMODULE += $(filter cpu_check_address, $(FEATURES_USED))
+
+# include periph_common if any periph_* driver is used
+ifneq (,$(filter periph_%, $(USEMODULE)))
+  USEMODULE += periph_common
+endif
+
+# set all USED periph_% init modules as DEFAULT_MODULE
+ifneq (,$(filter periph_init, $(USEMODULE)))
+    # To enable adding periph_% modules through the environment we cant used
+    # PERIPH_FEATURES since the MODULE might be have added directly as USEMODULE
+    PERIPH_MODULES_NO_INIT = periph_init% periph_common
+    # Use immediate expansion to get list of periph_% MODULEs without addition
+    # of periph_init_% MODULEs, use := since this is always expanded
+    PERIPH_MODULES := $(filter periph_%,$(USEMODULE))
+    PERIPH_INIT_MODULES := $(subst periph_,periph_init_,\
+      $(filter-out $(PERIPH_MODULES_NO_INIT),$(PERIPH_MODULES)))
+    DEFAULT_MODULE += $(PERIPH_INIT_MODULES)
+endif
+
+# add periph_init_% modules to USEMODULE unless disabled
+ifneq (,$(filter periph_init, $(USEMODULE)))
+  USEMODULE += $(filter $(PERIPH_INIT_MODULES),\
+    $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE)))
+endif
```

Comment:
I think just using

```C
USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))
```

would be easier to read. Double-adding default modules is no problem, as in `$(RIOTBASE)/Makefile.dep` the modules in `USEMODULE` are de-duplicated anyway.

Maybe it would make even more sense to just move the existing addition of default modules into `$(RIOTBASE)/Makefile.dep` to be part of the recursion? (And remove the already existing line adding them in favor of this?) Sooner or later there might be a use case to pull in optional module dependencies unless the user/application disables this. And this would allow `DEFAULT_MODULES` to be extended during normal dependency resolution.

Reply:
> ```C
> USEMODULE += $(filter-out $(DISABLE_MODULE),$(DEFAULT_MODULE))
> ```

I didn't do this because I wanted to make sure that not any module added to `DEFAULT_MODULE` after the first iteration of `Makefile.dep` is parsed.

> Maybe it would make even more sense to just move the existing addition of default modules into `$(RIOTBASE)/Makefile.dep` to be part of the recursion? (And remove the already existing line adding them in favor of this?) Sooner or later there might be a use case to pull in optional module dependencies unless the user/application disables this. And this would allow `DEFAULT_MODULES` to be extended during normal dependency resolution.

This can lead to some issues. Since I've had some discussions about this in other PRs as well I opened an issue on how to handle `DEAFAULT_MODULE` and how to limit the scope, I think it is better to reach a consensus there #13651.

Reply:
> I didn't do this because I wanted to make sure that not any module added to DEFAULT_MODULE after the first iteration of Makefile.dep is parsed.

I also wanted those module to show up if doing `make info-debug-vairable-DEFAULT_MODULE` makes it clear what modules can and cant be disabled.